diff -urN nspr-4.12.orig/nspr/configure.in nspr-4.12/nspr/configure.in
--- nspr-4.12.orig/nspr/configure.in	2016-05-14 19:11:06.693940400 +0200
+++ nspr-4.12/nspr/configure.in	2016-05-14 19:12:55.913195322 +0200
@@ -552,8 +552,8 @@
 dnl ========================================================
 dnl Checks for compilers.
 dnl ========================================================
-if test "$target" != "$host"; then
-    echo "cross compiling from $host to $target"
+if test "$target" != "$build"; then
+    echo "cross compiling from $build to $target"
     cross_compiling=yes
 
     case "$build:$target" in
@@ -582,7 +582,7 @@
 AC_PROG_CC
 
 dnl Reenter the conditional blocks after invoking AC_PROG_CC.
-if test "$target" != "$host"; then
+if test "$target" != "$build"; then
     if test -n "$USE_CPLUS"; then
         AC_CHECK_PROGS(CXX, $CXX "${target_alias}-g++" "${target}-g++", echo)
         unset ac_cv_prog_CXX
@@ -610,7 +610,7 @@
     _SAVE_CFLAGS="$CFLAGS"
     _SAVE_LDFLAGS="$LDFLAGS"
 
-    AC_MSG_CHECKING([for $host compiler])
+    AC_MSG_CHECKING([for $build compiler])
     AC_CHECK_PROGS(HOST_CC, $HOST_CC gcc cc /usr/ucb/cc, "")
     if test -z "$HOST_CC"; then
         AC_MSG_ERROR([no acceptable cc found in \$PATH])
@@ -621,10 +621,10 @@
     CFLAGS="$HOST_CFLAGS"
     LDFLAGS="$HOST_LDFLAGS"
 
-    AC_MSG_CHECKING([whether the $host compiler ($HOST_CC $HOST_CFLAGS $HOST_LDFLAGS) works])
+    AC_MSG_CHECKING([whether the $build compiler ($HOST_CC $HOST_CFLAGS $HOST_LDFLAGS) works])
     AC_TRY_COMPILE([], [return 0;],
         [AC_MSG_RESULT([yes])],
-        [AC_MSG_ERROR([installation or configuration problem: $host compiler $HOST_CC cannot create executables.])] )
+        [AC_MSG_ERROR([installation or configuration problem: $build compiler $HOST_CC cannot create executables.])] )
 
     CC=$_SAVE_CC
     CFLAGS=$_SAVE_CFLAGS
@@ -1163,9 +1163,9 @@
 fi
 
 dnl ========================================================
-dnl Override of system specific host options
+dnl Override of system specific build options
 dnl ========================================================
-case "$host" in
+case "$build" in
 *-mingw*|*-msys*)
     NSINSTALL=nsinstall
     ;;
